numba/mviewbuf.c

```diff
@@ -1,8 +1,22 @@
 #include "_pymodule.h"
 
-static int get_buffer(PyObject* obj, Py_buffer *buf)
+static int get_buffer(PyObject* obj, Py_buffer *buf, int force)
 {
-    return PyObject_GetBuffer(obj, buf, PyBUF_WRITABLE|PyBUF_ND|PyBUF_STRIDES|PyBUF_FORMAT);
+    Py_buffer read_buf;
+    int flags = PyBUF_ND|PyBUF_STRIDES|PyBUF_FORMAT;
+
+    /* Attempt to get a writable buffer */
+    if (!PyObject_GetBuffer(obj, buf, flags|PyBUF_WRITABLE))
+        return 0;
+    if (!force)
+        return -1;
+
+    /* Make a writable buffer from a read-only buffer */
+    PyErr_Clear();
+    if(-1 == PyObject_GetBuffer(obj, &read_buf, flags))
+        return -1;
+    return PyBuffer_FillInfo(buf, NULL, read_buf.buf, read_buf.len, 0,
```

Comment:
I tried that first, but it leads to an exception later on with the message "Expected a writable buffer interface" (to paraphrase) - I think at the time we hypothesised that something takes a new buffer (I guess it is able to because it knows the exporter of the buffer?) and fails because it can't get another writable one.

Can I safely release the read buffer at the point? I am concerned that if the read-buffer is released here, I don't know if something might be done to the buffer's data by NumPy whilst the writable buffer object's data pointer is still pointed at it.